<!DOCTYPE html>
<html lang="en-us">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <title>k-mer counting, part I: Introduction | BioInfoLogics</title>
    <link rel="stylesheet" href="/css/style.css" />
    <link rel="stylesheet" href="/css/fonts.css" />
    
  </head>

  <body>
    <nav>
    <ul class="menu">
      
      <li><a href="/">Home</a></li>
      
      <li><a href="/post/">Posts</a></li>
      
      <li><a href="/categories/">Categories</a></li>
      
      <li><a href="/tags/">Tags</a></li>
      
    </ul>
    <hr/>
    </nav>

<div class="article-meta">
<h1><span class="title">k-mer counting, part I: Introduction</span></h1>
<h2 class="author">Bernardo J. Clavijo</h2>
<h2 class="date">2018/09/17</h2>
</div>

<main>


<p>For the last 5 years, I have been advocating the analysis of <em>k-mer</em> spectra (in short, the frequencies at which different <em>k-mers</em> appear on different datasets) to evaluate genome assemblies. I have been using <em>k-mers</em> for many other analyses too, as most people doing Bioinformatics. And I also train others to perform <em>k-mer</em> analyses and have a long overdue debt to them to write down a more coherent version of the training. So here it goes, with the hope that it gets more people introduced to this simple and straightforward way to analyse sequence data.</p>

<p>This will be the first of a (yet undetermined) number of posts about k-mer counting on sequences, reads, assemblies, and more. They will be part conceptual, part practical. I will show practical bits using <a href="https://github.com/TGAC/KAT">KAT</a>, the k-mer analysis toolkit. Among other things, <a href="https://academic.oup.com/bioinformatics/article/33/4/574/2664339">KAT introduced the spectra comparisson as a way to assess genome assemblies vs WGS reads</a>, and we use it for all kinds of tricks.</p>

<p><br/><br/></p>

<hr />

<h1 id="k-mers-from-a-sequence"><em>k-mers</em> from a sequence</h1>

<p><a href="https://en.wikipedia.org/wiki/k-mer">What is a <em>k-mer</em> anyway?</a> It is a sequence of <em>k</em> characters in a string. So, by way of example, the sequence <code>ATCGATCAC</code> contains the following <em>3-mers</em>:</p>

<pre><code>Sequence: ATCGATCAC
3-mer #0: ATC
3-mer #1:  TCG
3-mer #2:   CGA
3-mer #3:    GAT
3-mer #4:     ATC
3-mer #5:      TCA
3-mer #6:       CAC
</code></pre>

<p>Or, in a nice looking table:</p>

<table>
<thead>
<tr>
<th align="center">Offset</th>
<th align="center">0</th>
<th align="center">1</th>
<th align="center">2</th>
<th align="center">3</th>
<th align="center">4</th>
<th align="center">5</th>
<th align="center">6</th>
</tr>
</thead>

<tbody>
<tr>
<td align="center"><strong>3-mer</strong></td>
<td align="center">ATC</td>
<td align="center">TCG</td>
<td align="center">CGA</td>
<td align="center">GAT</td>
<td align="center">ATC</td>
<td align="center">TCA</td>
<td align="center">CAC</td>
</tr>
</tbody>
</table>

<p>Obviously, we don&rsquo;t need to use <code>k=3</code> every time, so here you have the same example with <code>k=4</code>:</p>

<pre><code>Sequence: ATCGATCAC
4-mer #0: ATCG
4-mer #1:  TCGA
4-mer #2:   CGAT
4-mer #3:    GATC
4-mer #4:     ATCA
4-mer #5:      TCAC
</code></pre>

<p>Or, in a nice looking table:</p>

<table>
<thead>
<tr>
<th align="center">Offset</th>
<th align="center">0</th>
<th align="center">1</th>
<th align="center">2</th>
<th align="center">3</th>
<th align="center">4</th>
<th align="center">5</th>
</tr>
</thead>

<tbody>
<tr>
<td align="center"><strong>4-mer</strong></td>
<td align="center">ATCG</td>
<td align="center">TCGA</td>
<td align="center">CGAT</td>
<td align="center">GATC</td>
<td align="center">ATCA</td>
<td align="center">TCAC</td>
</tr>
</tbody>
</table>

<p>So far, so good, so easy. Any sequence of lenght <code>L</code> will contain <code>L - k + 1</code>   <em>k-mers</em> and you can easily imagine the many analyses that can be performed with these (if not, don&rsquo;t worry, I will show some along the way).</p>

<h3 id="did-you-know-images-did-you-know-png-why-are-k-mers-so-popular"><img src="/images/did_you_know.png" alt="did_you_know" /> <strong>Why are <em>k-mers</em> so popular ?</strong></h3>

<p>Decomposing a sequence into it&rsquo;s <em>k-mers</em> for analysis allows this set of fixed-size chunks to be analysed rather than the sequence, and this can be more efficient. K-mers are very useful in sequence matching (string matching with n-grams has a rich history), and set operations are faster, easier, and there is a lot of ready-available algorithms and techniques to work with them. A simple example: to check if a sequence <em>S</em> comes from organism <em>A</em> or from organism <em>B</em>, assuming the genomes of <em>A</em> and <em>B</em> are known and sufficiently different, we can check if S contains more <em>k-mers</em> present in <em>A</em> or in <em>B</em>. Yes, there are many tools that do just that.</p>

<h2 id="reverse-complement-and-canonical-k-mers">Reverse complement and canonical <em>k-mers</em></h2>

<p><small><em>*Note: for this section I assume double-strandness is the typical use case. This is a fair assumption, but there will be exceptions</em></small></p>

<p><em>K-mers</em> or <em><a href="https://en.wikipedia.org/wiki/n-gram">n-grams</a></em> have been around a long time, but a specific complication comes with bioinformatics: <a href="https://en.wikipedia.org/wiki/DNA#Base_pairing">DNA is normally double stranded, with bases paired on the oposite strands</a> and we normally read (or sequence) on any of the two strands. Still, we want to consider every location of the genome once no matter which strand we happened to have landed in.</p>

<p>In short: if we read the sequence <code>ATCGAC</code> that is an observation for that sequence <em>and it&rsquo;s reverse complement</em> <code>GTCGAT</code> to exist in the genome. One appears when reading the genome in one direction and the other on its oposite, we could have sequenced any of them. So for the sake of completeness we should perform all analyses by considering this sequence <code>ATCGAC/GTCGAT</code>.</p>

<p>Luckily, there is a way on <em>k-mer</em> analysis (or in sequence analysis in general) to be sure we always talk about the same sequence in the same way: we can only use the <em>canonical</em> sequence of a <em>k-mer</em> pair, this is the lexicographycally smaller of the two <em>reverse complementary</em> sequences. In other words, the one that comes before in alphabetical order.</p>

<p>Most <em>k-mer</em> counting tools will either count in canonical <em>k-mers</em> or have an option to switch between canonical and non-canonical. Most of the times canonical counting will be the default. This means if <code>GTCGAT</code> appears it will be counted as <code>ATCGAC</code>. Most of the times this is exactly what we want.</p>

<p>Here are the <em>3-mers</em> of <code>ATCGATCAC</code>, this time including the <em>reverse complement</em> and the <em>canonical</em> of each <em>3-mer</em> too.</p>

<table>
<thead>
<tr>
<th align="center">Offset</th>
<th align="center">0</th>
<th align="center">1</th>
<th align="center">2</th>
<th align="center">3</th>
<th align="center">4</th>
<th align="center">5</th>
<th align="center">6</th>
</tr>
</thead>

<tbody>
<tr>
<td align="center"><strong>3-mer</strong></td>
<td align="center">ATC</td>
<td align="center">TCG</td>
<td align="center">CGA</td>
<td align="center">GAT</td>
<td align="center">ATC</td>
<td align="center">TCA</td>
<td align="center">CAC</td>
</tr>

<tr>
<td align="center"><strong>Reverse Complement</strong></td>
<td align="center">GAT</td>
<td align="center">CGA</td>
<td align="center">TCG</td>
<td align="center">ATC</td>
<td align="center">GAT</td>
<td align="center">TGA</td>
<td align="center">GTG</td>
</tr>

<tr>
<td align="center"><strong>Canonical</strong></td>
<td align="center">ATC</td>
<td align="center">CGA</td>
<td align="center">CGA</td>
<td align="center">ATC</td>
<td align="center">ATC</td>
<td align="center">TCA</td>
<td align="center">CAC</td>
</tr>
</tbody>
</table>

<p>As you can see, in some cases the canonical <em>k-mer</em> appeared in the original sequence, but in some cases it was the <em>reverse complement</em>.</p>

<h3 id="did-you-know-images-did-you-know-png-why-do-many-tools-use-odd-values-for-k"><img src="/images/did_you_know.png" alt="did_you_know" /> Why do many tools use odd values for <em>k</em> ?</h3>

<p>To write a <em>k-mer</em> based tool, we need to account for the relationship within a <em>k-mer</em> and its <em>reverse complement</em> because we don&rsquo;t know which direction we are reading from. In many tools this relationship is used to detect sequence orientation: if sequence A has the same <em>k-mers</em> as sequence B, in oposite order and in <em>reverse complement</em>, B is in oposite orientation than A. But when using even values of k, a <em>k-mer</em> can be its own <em>reverse complement</em>, which makes it difficult to handle the <em>reverse complements</em>, and a lot less clear about orientations. This happens when the first half of the sequence is the <em>reverse complement</em> of the second half, as in   <code>TCGCGA</code>. A <em>k-mer</em> of odd <em>k</em> (an <em>odd-mer</em>, if you will) can never be its own reverse complement, because the base in the middle of its sequence should be its own complement. Hence, many tools stick to odd values for <em>k</em>.</p>

<h2 id="distinct-unique-and-total-k-mers">Distinct, unique, and total <em>k-mers</em></h2>

<p>People that counts <em>k-mers</em> talks a lot about unique <em>k-mers</em>, and there is also the <em>total</em> and the <em>distinct</em> counts to keep in mind. All of these can be also done in <em>canonical</em> or <em>non-canonical</em> ways. Sounds confusing? It is actually not.</p>

<p><em>Total k-mers</em> are the simplest of all, as we have shown with the example of <code>ATCGATCAC</code> the ammount of <em>k-mers</em> from a sequence varies with <em>k</em>, and that is basically it. <code>ATCGATCAC</code> has 7 <em>total 3-mers</em> and 6 <em>total 4-mers</em>.</p>

<p><em>Distinct k-mers</em> are counted only once, even if they appear more times, while <em>Unique k-mers</em> are those that appear only once. These two categories of <em>k-mers</em> are greatly affected by counting <em>canonically</em> or <em>not-canonically</em>, because <em>reverse complements</em> are being considered the same <em>k-mer</em>.</p>

<p>This is the <em>3-mer</em> composition for <code>ATCGATCAC</code> again, in <em>non-canonical</em> form, grouping its <em>k-mer</em> occurences:</p>

<table>
<thead>
<tr>
<th align="center"><em>k-mer</em></th>
<th align="center">Total count</th>
<th align="center">Distinct count</th>
<th align="center">Unique count</th>
</tr>
</thead>

<tbody>
<tr>
<td align="center">ATC</td>
<td align="center">2</td>
<td align="center">1</td>
<td align="center">0</td>
</tr>

<tr>
<td align="center">TCG</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">1</td>
</tr>

<tr>
<td align="center">CGA</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">1</td>
</tr>

<tr>
<td align="center">GAT</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">1</td>
</tr>

<tr>
<td align="center">TCA</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">1</td>
</tr>

<tr>
<td align="center">CAC</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">1</td>
</tr>

<tr>
<td align="center"><strong>Sum</strong></td>
<td align="center"><strong>7</strong></td>
<td align="center"><strong>6</strong></td>
<td align="center"><strong>5</strong></td>
</tr>
</tbody>
</table>

<p>And this is the <em>3-mer</em> composition for <code>ATCGATCAC</code> again, but in <em>canonical</em> form:</p>

<table>
<thead>
<tr>
<th align="center"><em>k-mer</em></th>
<th align="center">Total count</th>
<th align="center">Distinct count</th>
<th align="center">Unique count</th>
</tr>
</thead>

<tbody>
<tr>
<td align="center">ATC <small>(incl. GAT)</small></td>
<td align="center">3</td>
<td align="center">1</td>
<td align="center">0</td>
</tr>

<tr>
<td align="center">CGA <small>(incl. TCG)</small></td>
<td align="center">2</td>
<td align="center">1</td>
<td align="center">0</td>
</tr>

<tr>
<td align="center">TCA</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">1</td>
</tr>

<tr>
<td align="center">CAC</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">1</td>
</tr>

<tr>
<td align="center"><strong>Sum</strong></td>
<td align="center"><strong>7</strong></td>
<td align="center"><strong>4</strong></td>
<td align="center"><strong>2</strong></td>
</tr>
</tbody>
</table>

<p>As expected, in this <em>canonical</em> counting version TCG has been grouped into CGA, and GAT has been grouped into ATC. So the count has fewer <em>distinct k-mers</em>, and fewer <em>unique k-mers</em>. Canonical counting will always yield fewer (or at most the same), <em>distinct k-mers</em> and <em>unique k-mers</em>.</p>

<h3 id="did-you-know-images-did-you-know-png-what-makes-unique-k-mers-so-special"><img src="/images/did_you_know.png" alt="did_you_know" /> <strong>What makes unique <em>k-mers</em> so special ?</strong></h3>

<p>Unique k-mers make analyses simple. For instance, if a k-mer appears only once in a genome, it can be used to check if a particular region of the genome is present or not in a sample, or to check for variation, or as anchor to aligne other genome&rsquo;s content, or reads. Unique content within a genome is a good thing, enabling independent analysis of each part of the genome. Genes being usually made up of quite different content than the &ldquo;repeats&rdquo; of a genome is the reason why it is in general easier to assemble the genic part of a genome.</p>

<p>Sadly, repetitive content as a concept is poorly defined between biological and computational terms. While in biological terms a loose definition of repeat could be &ldquo;the stuff between the genes, which is assumed to be similar and low-complexity&rdquo;, in computational terms it always means &ldquo;a portion of sequence that appears exactly the same more than one time&rdquo;. This is why sometimes gene families (groups of very similar genes) are very hard to analyse as they appear multiple times in the genome. At the same time, this dissonance, also gives some light at the end of the tunnel in unexpected scenarios: a so-called biological repeat may turn out to have a few unique k-mers every time it appears, which in turn may make it a relatively simple region to analyse with the right choice of sequencing technologies and algorithms.</p>

<p><br/><br/></p>

<hr />

<h1 id="terminal-images-terminal-png-counting-k-mers-in-a-small-genome"><img src="/images/terminal.png" alt="terminal" /> Counting <em>k-mers</em> in a (small) genome</h1>

<p>We will start with an easy example first: the <a href="https://www.ncbi.nlm.nih.gov/nuccore/9626372">phi-X174 genome</a> has 5386 bp and is a simple non-repetitive genome.</p>

<p>We can use <code>kat hist</code> to count <em>27-mers</em> on the genome and check how many times each <em>27-mer</em> appears (we start with <code>k = 27</code> because KAT uses that as default):</p>

<pre><code class="language-sh">$ kat hist -o phiX.hist phiX.fasta
</code></pre>

<p>Checking the <code>phiX.hist</code> histogram (A.K.A. kmer spectrum) file, every <em>27-mer</em> in the genome appears only once. After the header lines starting with <code>#</code>, every line has a copy number (A.K.A. frequency) and a number of <em>k-mers</em>.</p>

<pre><code># Title:27-mer spectra for: phiX.fasta
# XLabel:27-mer frequency
# YLabel:# distinct 27-mers
# Kmer value:27
# Input 1:../genomes/phiX.fasta
###
1 5360
2 0
3 0
4 0
...
</code></pre>

<p>It is easy to see all <code>5360</code> <em>27-mers</em> appear once. Just to be sure we understand the logic, let&rsquo;s create a file where all the phi-X genome appears twice, count the <em>27-mers</em> and check the spectrum file.</p>

<pre><code>$ cat phiX.fasta phiX.fasta &gt; phiX_twice.fasta
$ kat hist -o phiX_twice.hist phiX_twice.fasta
</code></pre>

<pre><code># Title:27-mer spectra for: phiX_twice.fasta
# XLabel:27-mer frequency
# YLabel:# distinct 27-mers
# Kmer value:27
# Input 1:phiX_twice.fasta
###
1 0
2 5360
3 0
4 0
...
</code></pre>

<p>Now all 5360 27-mers appear twice. With the basics done, it is time to get into the most popular question on <em>k-mer</em> based analysis.</p>

<h3 id="choosing-k-the-size-of-the-universe-and-specificity-vs-sensitivity">Choosing <em>k</em>: the size of the universe and specificity vs. sensitivity</h3>

<p>Choosing <em>k</em> is not a trivial topic on <em>k-mer</em> analysis. Using a <em>k</em> that is too small will result on many unrelated sequences being composed of the same <em>k-mers</em>, in a textbook case of specificity loss because there being very few possible <em>k-mers</em> of that size. In the extreme of the small k, <code>k=1</code> only distinguishes two <em>canonical k-mers</em>: <code>A</code> and <code>C</code>. 1-mer analysis is incredibly popular in biology, but it is best known by the name of <em>GC content analysis</em>. On the other hand, using extremely large <em>k</em> values would sacrifice many of the benefits of <em>k-mer</em> analyses in the first place, such as decomposing the larger problem into small easy to solve exact problems. This leads to problems in inexact matching which I will describe in future posts.</p>

<p>Knowing how many different <em>k-mers</em> can be constructed at a certain <em>k</em> is a good starting point to make an informed choice of <em>k</em>. Since every of the <em>k</em> nucleotides in a <em>k-mer</em> can take any of the <code>A</code>, <code>C</code>, <code>G</code> or <code>T</code> values, the possible combinations of <em>k</em> positions are computed as <code>4^k</code>. When counting <em>canonical k-mers</em> not every value will be considered, but only those that are canonical. For odd values of <em>k</em> that is half of the <em>k-mers</em>, so the size of the universe becomes <code>(4^k)/2</code>. For even values, due to some <em>k-mers</em> being their own <em>reverse complement</em>, the size is slightly higher at <code>(4^k + 4^(k/2) ) / 2</code>. Again, sticking to odd <em>k</em> values makes life simpler.</p>

<p>Let&rsquo;s check now what happens if we change that <em>k</em> value between two carefully chosen values of 9 and 8 in the Phi-X genome (we will discuss how these were chosen soon). For historical reasons (due the to options from <a href="https://github.com/gmarcais/Jellyfish">Jellyfish</a>, on which KAT is based), KAT uses the <code>-m</code> parameter to set <em>k</em>. If <em>k</em> is small enough, the <em>k-mers</em> will appear more than once in the genome, even for this very simple genome.</p>

<p>So counting with <code>k=9</code> is done like this:</p>

<pre><code class="language-shell">$ kat hist -o phiX_9mer.hist -m 9 phiX.fasta
</code></pre>

<p>Then the <code>phiX_9mer.hist</code> file looks like this:</p>

<pre><code># Title:9-mer spectra for: phiX.fasta
# XLabel:9-mer frequency
# YLabel:# distinct 9-mers
# Kmer value:9
# Input 1:phiX.fasta
###
1 4972
2 189
3 8
4 1
5 0
6 0
7 0
8 0
9 0
...
</code></pre>

<p>Not every <em>9-mer</em> in the genome appear only once: there are <code>189</code> <em>9-mers</em> appearing twice, <code>8</code> <code>9-mers</code> appearing trice, and <code>1</code> <em>9-mer</em> that appears four times.</p>

<p>So, the number of <em>distinct 9-mers</em> in the phi-X genome is <code>4972 + 189 + 8 + 1 = 5170</code>, of these only <code>4972</code> are <em>unique 9-mers</em> (they only appear once in the genome), and the total number of <em>9-mers</em> is <code>4972 * 1 + 189 * 2 + 8 * 3 + 1 * 4 = 5378</code>.</p>

<p>The relationship between <em>total</em>, <em>distinct</em> and <em>unique</em> k-mers in a genome gives insight into its composition and the expected performance of different analyses, and depends on the choice of <em>k</em> , as illustrated by the effect of using <em>8-mers</em> on phi-X for the same analysis.</p>

<pre><code class="language-shell">$ kat hist -o phiX_8mer.hist -m 8 phiX.fasta
</code></pre>

<p>Now the histogram file looks like this:</p>

<pre><code># Title:8-mer spectra for: phiX.fasta
# XLabel:8-mer frequency
# YLabel:# distinct 8-mers
# Kmer value:8
# Input 1:phiX.fasta
###
1 4159
2 491
3 67
4 8
5 1
6 0
7 0
8 0
9 0
</code></pre>

<p>Here, only <code>4159</code> <em>8-mers</em> are <em>unique</em>, out of  <code>3547</code> <em>distinct 8-mers</em>, that are present in the genome&rsquo;s <code>5377</code> <em>total 8-mers</em>.</p>

<p><br/></p>

<p>The full response of the <em>unique</em>, <em>distinct</em> and <em>total</em> k-mer counts for phi-X can be computed by running successive KAT hist analyses, and then a bit of clever awk.</p>

<pre><code class="language-shell">#!/bin/bash
#Counting k-mers for different k on phiX
echo &quot;k,unique,distinct,total&quot;
for k in {1..15}; do
	kat hist -o phiX_$k.hist -m $k phiX.fasta &gt;/dev/null 2&gt;&amp;1
	egrep -v '^#' phiX_$k.hist|awk '{if ($1==1) u=$2; d+=$2; t+=$2*$1;}\
	END{print &quot;'$k',&quot;u&quot;,&quot;d&quot;,&quot;t}'
done

</code></pre>

<p>Executing this script gives a csv version of the following table:</p>

<table>
<thead>
<tr>
<th align="center">k</th>
<th align="center">unique</th>
<th align="center">distinct</th>
<th align="center">total</th>
</tr>
</thead>

<tbody>
<tr>
<td align="center">1</td>
<td align="center">0</td>
<td align="center">2</td>
<td align="center">5386</td>
</tr>

<tr>
<td align="center">2</td>
<td align="center">0</td>
<td align="center">10</td>
<td align="center">5385</td>
</tr>

<tr>
<td align="center">3</td>
<td align="center">0</td>
<td align="center">32</td>
<td align="center">5384</td>
</tr>

<tr>
<td align="center">4</td>
<td align="center">0</td>
<td align="center">135</td>
<td align="center">5383</td>
</tr>

<tr>
<td align="center">5</td>
<td align="center">10</td>
<td align="center">506</td>
<td align="center">5382</td>
</tr>

<tr>
<td align="center">6</td>
<td align="center">434</td>
<td align="center">1727</td>
<td align="center">5381</td>
</tr>

<tr>
<td align="center">7</td>
<td align="center">2295</td>
<td align="center">3547</td>
<td align="center">5380</td>
</tr>

<tr>
<td align="center">8</td>
<td align="center">4159</td>
<td align="center">4726</td>
<td align="center">5379</td>
</tr>

<tr>
<td align="center">9</td>
<td align="center">4972</td>
<td align="center">5170</td>
<td align="center">5378</td>
</tr>

<tr>
<td align="center">10</td>
<td align="center">5254</td>
<td align="center">5315</td>
<td align="center">5377</td>
</tr>

<tr>
<td align="center">11</td>
<td align="center">5346</td>
<td align="center">5361</td>
<td align="center">5376</td>
</tr>

<tr>
<td align="center">12</td>
<td align="center">5369</td>
<td align="center">5372</td>
<td align="center">5375</td>
</tr>

<tr>
<td align="center">13</td>
<td align="center">5374</td>
<td align="center">5374</td>
<td align="center">5374</td>
</tr>

<tr>
<td align="center">14</td>
<td align="center">5373</td>
<td align="center">5373</td>
<td align="center">5373</td>
</tr>

<tr>
<td align="center">15</td>
<td align="center">5372</td>
<td align="center">5372</td>
<td align="center">5372</td>
</tr>
</tbody>
</table>

<p>Which plotted, alongside the corresponding sizes of the universe is:</p>

<p><img src="/images/K001-phix_k_response.png" alt="phix_k_response" /></p>

<p>From these results, it is clear there is not a lot to look at for <code>k&lt;5</code>, these universes are so small they get completely saturated with the same <em>k-mers</em> over and over again. While frequency analyses aking to <em>GC content</em> and <em>codon usage</em> can be done there, it is slightly tangential to our topic today.</p>

<p>In the range <code>5 &lt;= k &lt;= 10</code>, there is a progressive growth on unique and distinct <em>k-mers</em>. Here, the <em>k</em> value makes a difference and will affect the results as the shown in the cherry-picked <em>9-mer</em> and <em>8-mer</em> cases earlier.</p>

<p>For <code>k &gt; 10</code>, pretty much every <em>k-mer</em> is <em>unique</em>. As a matter of fact for <code>k &gt;= 13</code>, every <em>k-mer</em> is a <em>unique k-mer</em>.</p>

<p><br></p>

<p>To make things more interesting, the same analysis can be done for the <a href="https://www.ncbi.nlm.nih.gov/nuccore/gi%7C556503834"><em>E. coli</em> genome</a>:</p>

<pre><code class="language-shell">#!/bin/bash
#Counting k-mers for different k on ecoli
echo &quot;k,unique,distinct,total&quot;
for k in {1..31}; do
	kat hist -o ecoli_$k.hist -m $k -h 3000000 ecoli.fasta &gt;/dev/null 2&gt;&amp;1
	egrep -v '^#| 0$' ecoli_$k.hist|awk '{if ($1==1) u=$2; d+=$2; t+=$2*$1;}\
	END{print &quot;'$k',&quot;0+u&quot;,&quot;d&quot;,&quot;t}'
done
</code></pre>

<p>And the corresponding plot looks like this:</p>

<p><img src="/images/K001-ecoli_k_response.png" alt="ecoli_k_response" /></p>

<p>We can notice two things straight away: in this larger, relatively more complex genomes there is high saturation of large universes up to <code>k=10</code> ; and the <em>unique</em> and <em>distinct</em> counts do not fully reach the <em>total k-mer</em> number, even well past after the desaturation. At <code>k=20</code> there are still some <code>k-mers</code> that are not unique: these are computational repeats.</p>

<p><br/>
<img src="/images/did_you_know.png" alt="did_you_know" /> <strong>Why do many tools use <em>k = 31</em> ?</strong></p>

<p>31 is the largest odd value of <em>k</em> such that a <em>k-mer</em> can be translated into a 64-bit integer. Most computers nowadays can manipulate 64-bit integers very well and most programming languages will make it easy to work efficiently with 64-bit integers (yes, this has to do with computers using 64-bit architectures). With <em>31-mers</em> being specific enough that a large number of <em>31-mers</em> both in mammalian-sized genomes (and some more complex genomes too) and in bacterial genome databases, bioinformatics is sold.</p>

<p><br/><br/></p>

<hr />

<p><strong>Comming soon&hellip;</strong></p>

<p>In future posts we will go over <em>k-mer</em> counting in WGS reads and the read spectrum, onwards to spectra comparisons, and all kinds of k-mer intersection based analyses.</p>

<hr />

<p><small></p>

<h3 id="version-history">Version History</h3>

<p>2018-09-17: initial post.</p>

<p></small></p>

</main>

  <footer>
  
<script type="application/javascript">
var doNotTrack = false;
if (!doNotTrack) {
	(function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
	(i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
	m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
	})(window,document,'script','https://www.google-analytics.com/analytics.js','ga');
	ga('create', 'UA-76944419-1', 'auto');
	
	ga('send', 'pageview');
}
</script>


  <div id="disqus_thread"></div>
<script type="application/javascript">
    var disqus_config = function () {
    
    
    
    };
    (function() {
        if (["localhost", "127.0.0.1"].indexOf(window.location.hostname) != -1) {
            document.getElementById('disqus_thread').innerHTML = 'Disqus comments not available by default when the website is previewed locally.';
            return;
        }
        var d = document, s = d.createElement('script'); s.async = true;
        s.src = '//' + "bioinfologics" + '.disqus.com/embed.js';
        s.setAttribute('data-timestamp', +new Date());
        (d.head || d.body).appendChild(s);
    })();
</script>
<noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
<a href="https://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>


  
  <hr/>
  &copy; <a href="http://earlham.ac.uk/bernardo-clavijo">Bernardo J. Clavijo</a> 2018 | <a href="https://github.com/bioinfologics">Github</a> | <a href="https://twitter.com/bjclavijo">Twitter</a>
  
  </footer>
  </body>
</html>

